sw/device/lib/dif/dif_i2c.c

```
@@ -0,0 +1,661 @@
+                                        .value = timing.data_signal_hold_time,
+                                    });
+  mmio_region_write32(i2c->base_addr, I2C_TIMING3_REG_OFFSET, timing3);
+
+  uint32_t timing4 = 0;
+  timing4 =
+      bitfield_set_field32(timing4, (bitfield_field32_t){
+                                        .mask = I2C_TIMING4_TSU_STO_MASK,
+                                        .index = I2C_TIMING4_TSU_STO_OFFSET,
+                                        .value = timing.stop_signal_setup_time,
+                                    });
+  timing4 =
+      bitfield_set_field32(timing4, (bitfield_field32_t){
+                                        .mask = I2C_TIMING4_T_BUF_MASK,
+                                        .index = I2C_TIMING4_T_BUF_OFFSET,
+                                        .value = timing.stop_signal_hold_time,
+                                    });
+  mmio_region_write32(i2c->base_addr, I2C_TIMING4_REG_OFFSET, timing4);
+}
+
+dif_i2c_result_t dif_i2c_init(mmio_region_t base_addr,
+                              dif_i2c_timing_params_t timing, dif_i2c_t *i2c) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  i2c->base_addr = base_addr;
+
+  // Turn the device off, before resetting it.
+  (void)dif_i2c_host_set_enabled(i2c, kDifI2cDisabled);
+
+  // Clear + disable all interrupts. We do this directly, rather than using the
+  // actual DIF functions, to perform both in single store operations.
+  mmio_region_write32(i2c->base_addr, I2C_INTR_STATE_REG_OFFSET, -1);
+  mmio_region_write32(i2c->base_addr, I2C_INTR_ENABLE_REG_OFFSET, 0);
+
+  // Reset all FIFO state. We do not check the errors for any of these
+  // functions, since we statically know that they cannot fail except for bad
+  // arguments, which we have already checked for.
+  (void)dif_i2c_reset_rx_fifo(i2c);
+  (void)dif_i2c_reset_fmt_fifo(i2c);
+  (void)dif_i2c_set_watermarks(i2c, kDifI2cLevel1Byte, kDifI2cLevel1Byte);
```

Don't discard errors.

---

Done.